### ISSUE TITLE
godot-sqlite fails to load on an m1 ARM MacBook Air

### ISSUE THREAD

**Issue by clayheaton** [2shady4u/godot-sqlite#47]
**Environment:**
 - OS: MacOS 11.2.3 (m1 MacBook Air)
 - Godot version: 3.3-rc9
 - _godot-sqlite_ version: 2.3

**Issue description:**
The SQLite Demo fails to run because the dylib is not suitable for the m1 ARM architecture on Macs.

**Steps to reproduce:**
Try to run the demo project in this repository.

**Minimal reproduction project:**
Use the demo project in this repository.

**Additional context**
I presume this is due to the project not being compiled for m1 Macs. I would be happy to compile and contribute the proper binary, but I will need a bit of help figuring out how to specify the proper architecture, etc.

Here are the errors when I try to run the demo project:

```
E 0:00:01.287   open_dynamic_library: Can't open dynamic library: /Users/clay/Documents/projects/godot/godot-sqlite/demo/addons/godot-sqlite/bin/osx/libgdsqlite.dylib, error: dlopen(/Users/clay/Documents/projects/godot/godot-sqlite/demo/addons/godot-sqlite/bin/osx/libgdsqlite.dylib, 2): no suitable image found.  Did find:
	/Users/clay/Documents/projects/godot/godot-sqlite/demo/addons/godot-sqlite/bin/osx/libgdsqlite.dylib: mach-o, but wrong architecture
	/Users/clay/Documents/projects/godot/godot-sqlite/demo/addons/godot-sqlite/bin/osx/libgdsqlite.dylib: mach-o, but wrong architecture.
  <C++ Error>   Condition "!p_library_handle" is true. Returned: ERR_CANT_OPEN
  <C++ Source>  platform/osx/os_osx.mm:1941 @ open_dynamic_library()

E 0:00:01.287   get_symbol: No valid library handle, can't get symbol from GDNative object
  <C++ Source>  modules/gdnative/gdnative.cpp:502 @ get_symbol()

E 0:00:01.287   init_library: No nativescript_init in "res://addons/godot-sqlite/bin/osx/libgdsqlite.dylib" found
  <C++ Source>  modules/gdnative/nativescript/nativescript.cpp:1479 @ init_library()
```

This manifests as:

"Attempt to call function 'new' in base 'NativeScript' on a null instance."

at line 73 of `database.gd`.

**Comment by 2shady4u** [comment]
Hello @clayheaton 

I'm reading a bit about M1 and it seems that Github Actions (the CI that I use for this project) doesn't maturely support M1 yet.
(https://github.com/actions/virtual-environments/issues/2187). There seems to be some possible work-arounds (like self-hosting), but nothing that seems to be easy to test on my end. 

If you want you can try to compile the binary yourself on your device (as described in [this](https://github.com/2shady4u/godot-sqlite#how-to-contribute) section of the README) and see if that fixes it. 

If that still generates errors, you might have to change the `-arch` flag on both [this](https://github.com/godotengine/godot-cpp/blob/d7c55b1ab2606253e94db5ed3caf75f90a47d03c/SConstruct#L220) and [this](https://github.com/2shady4u/godot-sqlite/blob/f4705a87e6ce914a3bdcd0ef631f79959eb6b61f/SConstruct#L220) to arm64e (or similar?).

I'll add official support as soon as Github Actions supports this, but atm my hands are tied I think 🤔

**Comment by 2shady4u** [comment]
Waiting for this PR to be merged: https://github.com/godotengine/godot-cpp/pull/529

**Comment by stebulba** [comment]
I have the same issues on Ubuntu 16.04 with Godot 3.3 stable.
Version 2.4 give me "Attempt to call function 'new' in base 'NativeScript' on a null instance." on line : SQLite.new()
on a different project, I use the version of Godot SQLite 1.5 and is Working.
Same when I import SQLite version 1.5 to my new project, it's working.
Until is fixe, maybe leave us older version.

**Comment by 2shady4u** [comment]
Hello @stebulba 

The issue described here is specific to the new ARM M1 architecture introduced by Apple for the new line of laptops and desktop computers.

In your case, the issue is related to the version of **glibc** which is == 2.23 in the case of Ubuntu 16.04 LTS.
The latest binaries of godot-sqlite are compiled using Ubuntu 20.04 LTS which uses glibc version 2.28
Legacy versions of godot-sqlite are available here and should work on 16.04:

If I may ask to try the following:
1. Find out the glibc version of machine, by running in the console/terminal:
`ldd --version`
so you can get the version of your glibc and report it back here. (It should be 2.23, but who knows!)

2. Replace the `libgdsqlite.so`-file by the legacy version (that supports older versions of glibc) as downloadable here:
https://github.com/2shady4u/godot-sqlite/actions/runs/819913518
Download the X11-legacy file and simply replace the `libgdsqlite.so`-file with this legacy file.

I might replace the linux binary file `libgdsqlite.so` with the legacy version in future versions of godot-sqlite as a lot of people seem to have issues with older versions of Linux.

**Comment by 2shady4u** [comment]
@clayheaton There have been some developments that might enable me to export the plugin to ARM64, namely: https://github.com/godotengine/godot-cpp/pull/584

I'm testing out some things on my testing repository (https://github.com/2shady4u/godot-cpp-ci) and I'll report back once I have some results.

**Comment by 2shady4u** [comment]
@clayheaton I've added a build for MacOS ARM64 to Github Actions which can be downloaded [here](https://github.com/2shady4u/godot-sqlite/actions/workflows/macos_builds.yml).

### Why is this build not included in Godot SQLite by default?
Currently Godot doesn't allow multiple MacOS architectures to be defined in the gdnlib-file, meaning that users without ARM64 wouldn't be able to use Godot SQlite anymore if this was the default. This might change once "universal" MacOS targets become available... whenever that is...

**Comment by clayheaton** [comment]
Thank you for your persistence on this, @2shady4u! Do you know if there's an open task on the Godot project for universal MacOS targets?

**Comment by 2shady4u** [comment]
@clayheaton There's an open PR (https://github.com/godotengine/godot-cpp/pull/529) on the `godot-cpp`-repo that should hopefully enable me to build binaries for the "universal" target.

Unfortunately there hasn't been much activity on this PR 😒 ...
Also a PR (https://github.com/godotengine/godot-cpp/pull/584) was merged independently that enables building for ARM64 , but the "universal" target was omitted in that case 🤷 .

**Comment by Methamane** [comment]
Hello @2shady4u, I'm not sure if it's related to the original issue but I'm also having problems with using the addon.

### Environment:
- OS: MacOS 12 (M1 Mac Mini)
- Godot version: 3.4.2
- godot-sqlite version: 3.0

### Issue Description:
When trying to initialize the wrapper SQLite.new() I'm getting "Attempt to call function 'new' in base 'NativeScript' on a null instance".

E 0:00:00.187   open_dynamic_library: Can't open dynamic library: /Users/mane/Documents/Projects/Godot/Moni/addons/godot-sqlite/bin/osx/libgdsqlite.dylib, error: dlopen(/Users/mane/Documents/Projects/Godot/Moni/addons/godot-sqlite/bin/osx/libgdsqlite.dylib, 0x0002): tried: '/Users/mane/Documents/Projects/Godot/Moni/addons/godot-sqlite/bin/osx/libgdsqlite.dylib' (mach-o file, but is an incompatible architecture (have 'x86_64', need 'arm64e')), '/usr/lib/libgdsqlite.dylib' (no such file).
  <C++ Error>   Condition "!p_library_handle" is true. Returned: ERR_CANT_OPEN
  <C++ Source>  platform/osx/os_osx.mm:1917 @ open_dynamic_library()

Please let me know if I should open a separate issue or if additional info is needed.

**Comment by 2shady4u** [comment]
> Hello @2shady4u, I'm not sure if it's related to the original issue but I'm also having problems with using the addon.
> ### Environment:
> 
>     * OS: MacOS 12 (M1 Mac Mini)
> 
>     * Godot version: 3.4.2
> 
>     * godot-sqlite version: 3.0
> 
> 
> ### Issue Description:
> 
> When trying to initialize the wrapper SQLite.new() I'm getting "Attempt to call function 'new' in base 'NativeScript' on a null instance".
> 
> E 0:00:00.187 open_dynamic_library: Can't open dynamic library: /Users/mane/Documents/Projects/Godot/Moni/addons/godot-sqlite/bin/osx/libgdsqlite.dylib, error: dlopen(/Users/mane/Documents/Projects/Godot/Moni/addons/godot-sqlite/bin/osx/libgdsqlite.dylib, 0x0002): tried: '/Users/mane/Documents/Projects/Godot/Moni/addons/godot-sqlite/bin/osx/libgdsqlite.dylib' (mach-o file, but is an incompatible architecture (have 'x86_64', need 'arm64e')), '/usr/lib/libgdsqlite.dylib' (no such file). <C++ Error> Condition "!p_library_handle" is true. Returned: ERR_CANT_OPEN <C++ Source> platform/osx/os_osx.mm:1917 @ open_dynamic_library()
> 
> Please let me know if I should open a separate issue or if additional info is needed.

Hey @Methamane! I've just a released a new version of the plugin that exports the plugin to the universal MacOS target which should work regardless of your device.

You can download the release from the releases tab (https://github.com/2shady4u/godot-sqlite/releases/tag/v3.1) or from the Godot asset Lib.
Please tell me if this solves the issue 🙏

**Comment by Methamane** [comment]
> > Hello @2shady4u, I'm not sure if it's related to the original issue but I'm also having problems with using the addon.
> > ### Environment:
> > ```
> > * OS: MacOS 12 (M1 Mac Mini)
> > 
> > * Godot version: 3.4.2
> > 
> > * godot-sqlite version: 3.0
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ### Issue Description:
> > When trying to initialize the wrapper SQLite.new() I'm getting "Attempt to call function 'new' in base 'NativeScript' on a null instance".
> > E 0:00:00.187 open_dynamic_library: Can't open dynamic library: /Users/mane/Documents/Projects/Godot/Moni/addons/godot-sqlite/bin/osx/libgdsqlite.dylib, error: dlopen(/Users/mane/Documents/Projects/Godot/Moni/addons/godot-sqlite/bin/osx/libgdsqlite.dylib, 0x0002): tried: '/Users/mane/Documents/Projects/Godot/Moni/addons/godot-sqlite/bin/osx/libgdsqlite.dylib' (mach-o file, but is an incompatible architecture (have 'x86_64', need 'arm64e')), '/usr/lib/libgdsqlite.dylib' (no such file). <C++ Error> Condition "!p_library_handle" is true. Returned: ERR_CANT_OPEN <C++ Source> platform/osx/os_osx.mm:1917 @ open_dynamic_library()
> > Please let me know if I should open a separate issue or if additional info is needed.
> 
> Hey @Methamane! I've just a released a new version of the plugin that exports the plugin to the universal MacOS target which should work regardless of your device. You can download the release from the releases tab (https://github.com/2shady4u/godot-sqlite/releases/tag/v3.1) or from the Godot Asset Lib.
> 
> Please tell me if this solves the issue 🙏

Hey @2shady4u,
Thanks for the update! 
I didn't get a chance to do a more in depth testing but from what I'm seeing so far it's working great!

**Comment by 2shady4u** [comment]
Closed #47 due to inactivity

**Comment by BruceABeitman** [comment]
Hey, FYI I found that the 3.1 version **does** work for my M1, but surprisingly the 3.4 does not unfortunately.
